### PR TITLE
WS-2675/WS-2704: Scroll depth tracker for Article & Home page

### PR DESCRIPTION
**Pull request by LilyL0u**
Resolves JIRA: WS-2675 & WS-2704

This pull request introduces a new scroll depth tracking hook and integrates it into the `ArticlePage` component to enable analytics on how far users scroll through articles. The main focus is on tracking user engagement by reporting when users reach certain scroll thresholds, but only when the "Continue Reading" button is not shown or all content is visible.

**Scroll Depth Tracking Integration:**

- **New Hook Implementation:**
  - Added a new hook, `useScrollDepthTracker`, which tracks how far users scroll through a specified element and dispatches analytics events at 25%, 50%, 75%, and 100% scroll depths. The hook ensures tracking only occurs when analytics are enabled and the page is not in AMP or "Lite" mode. (`src/app/hooks/useScrollDepthTracker/index.tsx`)

- **Article Page Integration:**
  - Imported and initialized the `useScrollDepthTracker` hook in `ArticlePage.tsx`, setting it up to track the main article content. The hook is enabled only if the "Continue Reading" button is not shown or all content is visible, to avoid double tracking. (`src/app/pages/ArticlePage/ArticlePage.tsx`) [[1]](diffhunk://#diff-7b748265dea863c1d16b9d5339803ce5a801ecbb8584ae09860827e0de3a6848R64) [[2]](diffhunk://#diff-7b748265dea863c1d16b9d5339803ce5a801ecbb8584ae09860827e0de3a6848R370-R375)
  - Attached the hook's ref to the `<main>` element containing the article content, so scroll depth is measured relative to the main article. (`src/app/pages/ArticlePage/ArticlePage.tsx`)

requestAnimationFrame is a browser API that lets you schedule a function to run just before the browser repaints the screen. Instead of running the code immediately when a scroll event fires, you give it to the browser and tell it to run it on the next frame. The browser then calls it at a consistent rate (about 60 times per second) in sync with its own rendering cycle.  Using this means we don't recalculate scroll depth on every single scrolle event, just once per frame.
In the previous protorype, handleScroll ran its depth check synchonously on EVERY scrollevent, potentially hundreds of times per second as a user scrolls. Callin once per frame is less intensive.

sentThresholds: a Set that remembers which % values have already fired 
tickingRef: a flag to prevent running checkDepth more than once per animation frame
mainElement: the article DOM node (passed in via the ref callback at the bottom)
shouldTrack: false on AMP/Lite pages, when the feature toggle is off, or when `enabled=false`

In the tests, imagine the article is 1000px tall, and the viewport is 100px tall. As the user scrolls down, the bottom of edge of the window moves further down the article.

`scrollY=0:    window covers 0px–100px    → bottom at 100px  → 10% of article visible
scrollY=150:  window covers 150px–250px  → bottom at 250px  → 25% of article reached  ← fires 25% event
scrollY=400:  window covers 400px–500px  → bottom at 500px  → 50% of article reached  ← fires 50% event
scrollY=650:  window covers 650px–750px  → bottom at 750px  → 75% reached
scrollY=900:  window covers 900px–1000px → bottom at 1000px → 100% reached`

The numbers here are for ease of understanding in the tests, but any number can be used in the calculation. The calculation for the depth is 
`((scrollY + viewportHeight)) / articleHeight X 100`
e.g ((150+100)/1000) X 100 = 25%

the beforeAll sets mocked scrollY so that simulatScroll(150) sets it to 150.
jsdom sets innerHeight to 0 by default, there is no real browser window, so setting it to 100 to start with is like a 100px viewport.

cb(0) is a timestamp that represents number of milliseconds since page load, (cb is a conventional abbreviation for callback) and 0 means now.

Summary
======
_A very high-level summary of easily-reproducible changes that can be understood by non-devs, and why these changes where made._

Code changes
======
- _List key code changes that have been made._

Testing
======
1. _List the steps required to test this PR._

## Useful Links
 - [Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)
 - [Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)
- _Add Links to useful resources related to this PR if applicable._

### CODE REVIEW

File: src/app/hooks/useScrollDepthTracker/index.tsx [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+  const viewportBottom = window.scrollY + window.innerHeight;
+  const scrolledPast = viewportBottom - trackingStartY;
+  return Math.min(100, Math.max(0, (scrolledPast / trackingHeight) * 100));
+};
+
+const useScrollDepthTracker = (componentName: string, enabled = true) => {
+  const { isAmp, isLite } = use(RequestContext);
+  const { service } = use(ServiceContext);
+  const { trackingIsEnabled } = useTrackingToggle(componentName); // this is in the togglrd config to enable/disable trackong across the site
```
[the remark this answers is not in the document]
Comment (LilyL0u) [reply src/app/hooks/useScrollDepthTracker/index.tsx]: lol

File: src/app/hooks/useScrollDepthTracker/index.tsx [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      tickingRef.current = false;
+    };
+
+    const handleScroll = () => {
+      if (tickingRef.current) return;
+      tickingRef.current = true;
+      requestAnimationFrame(checkDepth);
+    };
+
+    handleScroll();
```
Comment (pvaliani) [inline src/app/hooks/useScrollDepthTracker/index.tsx:148]: This might send a scroll-depth event as soon as the page loads, before the normal page-view analytics has finished. That can make the page-view data wrong or incomplete. Might need to wait until page view has safely finished, or make sure scroll-depth events do not reset the page analytics

File: src/app/pages/ArticlePage/ArticlePage.tsx [hunk 3]
```diff
@@ -354,6 +355,12 @@ const ArticlePage = ({
     continueReadingButtonToggle,
   );
 
+  const scrollDepthEnabled = !showContinueReadingButton || showAllContent;
```
Comment (pvaliani) [inline src/app/pages/ArticlePage/ArticlePage.tsx]: This switches scroll tracking off whenever the article has a Continue Reading block. On desktop, that button is hidden and the full article is already visible, so desktop users on those articles may never send scroll-depth events. Might need to be tighter than just whether the block exists

File: src/app/hooks/useScrollDepthTracker/index.tsx [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+import { ServiceContext } from '#app/contexts/ServiceContext';
+import dispatchTrackingRequests from '#app/lib/analyticsUtils/dispatchTrackingRequests';
+import useTrackingToggle from '../useTrackingToggle';
+
+const SCROLL_DEPTH_THRESHOLDS = [25, 50, 75, 100] as const;
+
+type ScrollDepthThreshold = (typeof SCROLL_DEPTH_THRESHOLDS)[number];
+
+const getScrollDepthPercent = (element: HTMLElement) => {
+  const heroFigure = element.querySelector('figure');
```
Comment (pvaliani) [inline src/app/hooks/useScrollDepthTracker/index.tsx:26]: Think this just picks the first <figure> anywhere in the article and treats it like the hero image. If an article has text before its first inline image, we would ignore that earlier reading and start tracking too late

File: src/app/hooks/useScrollDepthTracker/index.tsx (hunk 2 above)
Comment (LilyL0u) [reply src/app/hooks/useScrollDepthTracker/index.tsx:148]: See home page PR for info on not resetting the page analytics.

Comment (pvaliani) [reply src/app/hooks/useScrollDepthTracker/index.tsx:148]: Ah my bad, I may have missed the context as I reviewed this one first!

File: src/app/pages/ArticlePage/ArticlePage.tsx (hunk 3 above)
Comment (LilyL0u) [reply src/app/pages/ArticlePage/ArticlePage.tsx]: Good point. I didn't test on a service with no button! Just checked, and it doesn't work. Will look into a fix.

Comment (pvaliani) [reply src/app/pages/ArticlePage/ArticlePage.tsx]: All good, nice one!

Comment (LilyL0u) [reply src/app/pages/ArticlePage/ArticlePage.tsx]: Fixed by adding a media query so that on desktop, the tracking is not effected by the presence of the button:


Desktop (1008px+): [isDesktopViewport](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) is true → scroll depth tracking is available immediately ✓
Mobile/tablet without button: [!showContinueReadingButton](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) is true → scroll depth is available immediately ✓
Mobile/tablet with button (collapsed): All conditions false → scroll depth disabled ✓
Mobile/tablet with button (expanded): [showAllContent](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) is true → scroll depth fires after click ✓
The behavior now syncs perfectly with the CSS breakpoints, so scroll depth tracking will be consistent with whether content is actually visible to the user.

File: src/app/hooks/useScrollDepthTracker/index.tsx (hunk 4 above)
Comment (LilyL0u) [reply src/app/hooks/useScrollDepthTracker/index.tsx:26]: I have looked into accommodating this edge case but I am not sure it is worth it for the extra complexity. The articles that have text before the hero image are so rare that I don't remember ever seeing any in the wild. Only on test.
